Added post package hook

GermanHydrogen opened the pull request:
I found the idea of a post-package hook  (https://github.com/csiro-fair/marimba/issues/13#issuecomment-2739179097)  useful, regardless of how #13 and #14 are ultimately resolved.
Since the package command records the checksum of each file in a dataset, the post package hook cannot really run after the usual package command if the code in the hook is to be allowed to modify files. I decided that it would be most useful to run the hook after the metadata file has been created. It also receives the root directory of the dataset (not the data directory), since the code in the hook should be able to modify the metadata file.

I have not yet modified the documentation (`docs/pipeline.md`), but am willing to do so if the general implementation is accepted.

This feature addition should not break the behavior of marimba.

cjackett commented:
Hi @GermanHydrogen,

Thank you for making a good start on this feature. While we haven't needed it just yet, we suspect that there will eventually be scenarios where we want to make surgical changes to any file that is packaged in the final dataset. This would require the post package hook to be at step 11/12 - right before the final package validation. 

Some scenarios where this would be useful include:
* Removing sensitive strings from the logs or pipeline code for anonymisation/privacy/security
* Extracting additional details after the collections are merged during packaging, and generating custom packaged dataset summaries and/or plots
* Having multiple separate pipelines that get packaged together and then needing to produce some derived product
* Creating dynamic documentation based on the final content of the packaged dataset
* I am sure there will be additional use cases that will arise in the future

This would, of course, change the hashes of any file that is changed, and we had thought that this could be managed by the interface design of the post package hook. The method implementation in BasePipeline could enforce a return type of a list of changed files paths, and the post package hook in core Marimba could iterate through those files and update the manifest, accounting for changes to any files within the packaged dataset. This would provide extraordinary flexibility, and I think it's only a matter of time until we need something like this.

That is all a bit more involved than probably what you need right now, but keeping in mind that you can add any code into the tail of your Pipeline `_package` method that can do extra things that Marimba doesn't do. For instance, here is an example where we create multi-level iFDO and summary files (at multiple levels in the data directory hierarchy) by just adding custom code at the end of the Pipeline `_package` method: https://github.com/csiro-fair/flowcam-pipeline/blob/19c61e3272b3a5a1920c3a0ab5f34f6ce168b8bb/flowcam.pipeline.py#L448

Would an approach like that solve your current use case?

GermanHydrogen commented:
Hi @cjackett,
I implemented your suggestions to move the post-package hook execution to the penultimate step of the package command and added a mechanism to update the manifest file after the post-package hook execution based on a set of file paths returned by the post-package hooks. I hope this covers your requirements.

In theory, your suggested workaround would also work for us, but I would prefer a long term solution through the post-package hook, as generating the metadata "manually" would result in the pipeline depending on marimba internals. As we currently need the solution only until #13 and #14 are resolved, having the "hot fix" contained in the post-package hook method would also simplify removing it in the end.

GermanHydrogen commented:
Hi @cjackett,
I rebased my branch off main and added documentation to the `_post_package` method in the `docs/pipeline.md` documentation.

GermanHydrogen commented:
Hi @cjackett,
I found a bug in the path resolution in the manifest generation when `marimba package` is used together with the `--project-dir` flag. This causes the the manifest generation to fail silently (errors are only output to dataset.log).

I thought the bug was caused by my changes, but later found out that the bug is also in the main branch. Let me know if you want a separate PR for this.

cjackett reviewed (approved):
Hi @GermanHydrogen,

Thank you for this PR implementing the post-package hook functionality. The implementation is well-designed and addresses an important need for modifying datasets after they're packaged but before validation.

I've tested the implementation and have made some additional changes to the documentation, including some extra examples to further demonstrate the key use cases, such as properly handling file renames with tracking of both source and destination paths, emphasising the importance of returning every changed file path to ensure manifest integrity during validation.

I verified the implementation with some real-world testing:
• I used a test pipline to test removing .env files
• Confirmed correct tracking of both source and destination paths during file renames
• Verified proper hash updating when modifying README.md content
• Validated that manifest validation works correctly after all modifications

It was a good pickup on the `--project-dir` fix - we tend to exclusively operate marimba from within a marimba project directory and so that option isn't a well-travelled logic path. Thank you.

This post-package hook will be particularly valuable for ensuring sensitive information is removed before dataset publication and for making final adjustments to files. Happy to approve this PR and merge this into the main branch.
